pinot-tools/src/main/java/com/linkedin/pinot/tools/perf/QueryRunner.java

```diff
@@ -0,0 +1,586 @@
+    final AtomicLong totalBrokerTime = new AtomicLong(0L);
+    final AtomicLong totalClientTime = new AtomicLong(0L);
+    final DescriptiveStatistics clientTimeStatistics = new DescriptiveStatistics();
+
+    ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+    for (int i = 0; i < numThreads; i++) {
+      executorService.submit(new Runnable() {
+        @Override
+        public void run() {
+          while (true) {
+            String query = queries.get(random.nextInt(numQueries));
+            try {
+              executeQueryInMultiThreads(driver, query, numQueriesExecuted, totalBrokerTime, totalClientTime,
+                  clientTimeStatistics);
+            } catch (Exception e) {
+              LOGGER.error("Caught exception while running query: {}", query, e);
+              return;
+            }
+          }
+        }
+      });
+    }
+    executorService.shutdown();
+
+    long startTime = System.currentTimeMillis();
+    int numReportIntervals = 0;
+    while (true) {
+      if (executorService.isTerminated()) {
+        LOGGER.error("All threads got exception and already dead.");
+        return;
+      }
+      Thread.sleep(reportIntervalMs);
+      long timePassed = System.currentTimeMillis() - startTime;
+      LOGGER.info("Time Passed: {}ms, Average QPS: {}, Average Broker Time: {}ms, Average Client Time: {}ms.",
+          timePassed, numQueriesExecuted.get() / ((double) timePassed / MILLIS_PER_SECOND),
+          totalBrokerTime.get() / (double) numQueriesExecuted.get(),
+          totalClientTime.get() / (double) numQueriesExecuted.get());
+      numReportIntervals++;
+      if (numReportIntervals == numIntervalsToReportClientTimeStatistics) {
+        reportClientTimeStatistics(clientTimeStatistics);
+        numReportIntervals = 0;
+      }
+    }
+  }
+
+  /**
+   * Use multiple threads to run query at a target QPS.
+   * <p>Use a concurrent linked queue to buffer the queries to be sent. Use the main thread to insert queries into the
+   * queue at the target QPS, and start <code>numThreads</code> worker threads to fetch queries from the queue and send
+   * them.
+   * <p>The main thread is responsible for collecting and logging the statistic information periodically.
+   * <p>Queries are picked randomly from the query file.
+   * <p>Query runner will run forever or until too many queries blocked in the queue.
+   *
+   * @param conf perf benchmark driver config.
+   * @param queryFile query file.
+   * @param numThreads number of threads sending queries.
+   * @param startQPS start QPS (target QPS).
+   * @param reportIntervalMs report interval in milliseconds.
+   * @param numIntervalsToReportClientTimeStatistics number of intervals to report detailed client time statistics.
+   * @throws Exception
+   */
+  @SuppressWarnings("InfiniteLoopStatement")
+  public static void targetQPSQueryRunner(PerfBenchmarkDriverConf conf, String queryFile, int numThreads,
```

Comment:
Any guideline for specifying number of threads for a given QPS? If so, please put in the comments.